### PR TITLE
feat: allow for persistence of config data

### PR DESCRIPTION
**Pull request by rsmitty**
This PR will allow users to set the `persist: true` value in their
config data to tell talos not to re-pull the config data at each reboot.
The default will still remain as a "pull every time" methodology in order
to encourage immutability by default.

Signed-off-by: Spencer Smith <robertspencersmith@gmail.com>

### CODE REVIEW

**Comment by smira** [comment]
I wonder if we want to umount `/boot` at some early point in boot to make sure it's "clean" on reboot? or mount/re-mount it as ro/rw as needed, basically my point is to make it either mounted only when we need it or make it rw only for the time we need to write config to minimize the risk of corruption, as it's FAT.

**Comment by smira** [comment]
And if we want to test it out, we can make it (an option? default?) for `osctl cluster create` which will go to integration tests... we can even stop config HTTP service after some reboots (or after successful install?). This should be optional to keep upgrade tests happy (as they need to be able to provision older clusters)

File: internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go [hunk 1]
```diff
@@ -135,6 +154,10 @@ func (d *Sequencer) Boot() error {
 			configtask.NewExtraFilesTask(),
 			configtask.NewSysctlsTask(),
 		),
+		phase.NewPhase(
```
Comment (rsmitty) [inline internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: Leaving this mounted as RO at the end here because other sequences like Upgrade want to read this file immediately when they start running.

File: internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go [hunk 2]
```diff
@@ -152,17 +175,7 @@ func (d *Sequencer) Boot() error {
...
-		return err
-	}
-
-	config, err := config.New(content)
-	if err != nil {
-		return err
-	}
-
-	phaserunner, err := phase.NewRunner(config, runtime.Shutdown)
+	phaserunner, err := phase.NewRunner(nil, runtime.Shutdown)
```
Comment (rsmitty) [inline internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: There was no need for shutdown to read the config file from what I could tell.

Comment (smira) [reply internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: what if we need to access it in the shutdown sequence some time in the future? if we get to shutdown phase, we should have got through the boot phase successfully, so config should be in `/boot`?

File: internal/app/machined/internal/phase/config/config.go [hunk 3]
```diff
@@ -30,30 +37,95 @@ func (task *Task) TaskFunc(mode runtime.Mode) phase.TaskFunc {
...
+
+			return nil
+		}
+
+		return err
+	}
+
+	log.Printf("found an existing config.yaml. determining if we should pull anyways")
+
+	content, err := config.FromFile(constants.ConfigPath)
```
Comment (smira) [inline internal/app/machined/internal/phase/config/config.go]: this might be _theoretical_, but still: what if config gets corrupted/it is unreadable for some reason, or due to some bug we got it saved in a wrong way?

before this PR, as config was fetched every time, this wasn't noticeable

after this PR:

* if persist is enabled, probably there's not much we can do (if original config source is gone)
* but if persist is not enabled, it means that node won't boot anymore, while technically there's no good reason to fail the boot

so proposal:

if config can't be read, treat it as non-fatal error and fetch config, so in other words unify case above then config file doesn't exist with case below when config file can't be read/parsed or config has !.Persist. 

Basically, try reading config, if that fails, log an error, and treat that as !Persist.

Comment (rsmitty) [reply internal/app/machined/internal/phase/config/config.go]: @smira reworked this to accomplish what you were talking about. if we have any issue either reading in the config from disk or parsing it, we'll force the re-fetch.

File: internal/app/machined/internal/phase/config/config.go [hunk 4]
```diff
@@ -30,30 +37,101 @@ func (task *Task) TaskFunc(mode runtime.Mode) phase.TaskFunc {
...
+	forceCfgPull := false
+
+	content, err := config.FromFile(constants.ConfigPath)
+	if err != nil {
+		forceCfgPull = true
+
+		log.Printf("failed to read config, forcing a re-pull of the config: %s", err)
+	}
+
+	cfg, err := config.New(content)
```
Comment (smira) [inline internal/app/machined/internal/phase/config/config.go]: I think this should go under `else` for the `if err != nil` above, as `content` will be `nil` if we can't read the file

File: internal/app/machined/internal/phase/config/config.go (hunk 3 above)
Comment (smira) [reply internal/app/machined/internal/phase/config/config.go]: thanks, this looks much better, but I feel we could try to unify it a bit more potentially simplifying a code.

if we had a function `readConfig() (runtime.Configurator, error)` which tries to read config, but returns an error if it's missing, we can call it, and if it returns an error or `!cfg.Persist()` try to fetch it. this would unify the code above (which checks if file exists and below which checks for readability). I think this will be less code, and we could have used this function in the sequencer as well as it seems to be doing same thing (reading config from file) over and over again?

File: internal/app/machined/internal/phase/config/config.go (hunk 4 above)
Comment (rsmitty) [reply internal/app/machined/internal/phase/config/config.go]: yep, good spot. fixed in amended commit.

File: internal/app/machined/internal/phase/config/config.go (hunk 3 above)
Comment (smira) [reply internal/app/machined/internal/phase/config/config.go]: so my point is that the case when config doesn't exist or when it's not readable is the same case

File: internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go (hunk 2 above)
Comment (rsmitty) [reply internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: fixed this in amended commit

File: pkg/config/config.go [hunk 5]
```diff
@@ -122,8 +122,8 @@ func NewConfigBundle(opts ...BundleOption) (*v1alpha1.ConfigBundle, error) {
 	return bundle, nil
 }
 
-// New initializes and returns a Configurator.
-func New(c Content) (config runtime.Configurator, err error) {
+// newConfig initializes and returns a Configurator.
```
Comment (rsmitty) [inline pkg/config/config.go]: went ahead and made this function, `fromFile`, and `fromBytes` private since I removed all references to them and replaced with the proper `NewConfigFrom*` function.

File: docs/website/content/v0.3/en/configuration/v1alpha1.md [hunk 6]
```diff
@@ -72,6 +72,19 @@ Valid Values:
 
 - ``v1alpha1``
 
+#### persist
+
+Indicates whether to pull the config data upon every boot.
```
Comment (andrewrynhard) [inline docs/website/content/v0.3/en/configuration/v1alpha1.md]: nit: We refer to this config as "machine config" in the rest of the docs.

File: pkg/config/config.go (hunk 5 above)
Comment (andrewrynhard) [reply pkg/config/config.go]: nit: The package is already called "`config`". The extra "`Config`" in the `NewConfigFrom*` functions is redundant.

File: pkg/constants/constants.go [hunk 7]
```diff
@@ -168,7 +168,7 @@ const (
 	EtcdDataPath = "/var/lib/etcd"
 
 	// ConfigPath is the path to the downloaded config.
-	ConfigPath = "/run/config.yaml"
+	ConfigPath = "/boot/config.yaml"
```
Comment (andrewrynhard) [inline pkg/constants/constants.go]: A note for my future self: I wonder if we should call this `machine.yaml` since we always refer to it as "machine config".

File: internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go [hunk 8]
```diff
@@ -122,10 +124,26 @@ func (d *Sequencer) Boot() error {
...
 			rootfs.NewVarDirectoriesTask(),
 		),
+		phase.NewPhase(
+			"unmount /boot",
+			rootfs.NewUnmountSystemDisksTask(constants.BootPartitionLabel),
+		),
+		phase.NewPhase(
+			"mount /boot as rw",
+			rootfs.NewMountSystemDisksTask(constants.BootPartitionLabel),
+		),
```
Comment (andrewrynhard) [inline internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: Should we maybe mount as R/W to begin with earlier in boot? Then we won't have to do these.

Comment (rsmitty) [reply internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: Andrey's concern around that was that because it's a FAT file system there's a higher chance of corruption. So we should mount as rw only when we needed it.

**Comment by rsmitty** [comment]
@andrewrynhard fixed your nits in amended commit.

File: internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go (hunk 8 above)
Comment (smira) [reply internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go]: my concern was close to that, but basically not to leave it r/w longer than we need, as if we screw `/boot`, we screw whole node
